pkg/declarative/v2/reconciler.go

```diff
@@ -0,0 +1,382 @@
+)
+
+type ConditionReason string
+
+const (
+	ConditionReasonResourcesAreAvailable ConditionReason = "ResourcesAvailable"
+	ConditionReasonReady                 ConditionReason = "Ready"
+)
+
+//nolint:funlen,nestif,gocognit,gocyclo,cyclop,maintidx //TODO discuss if worth breaking up or if more readable as is.
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	obj := r.prototype.DeepCopyObject().(Object)
+
+	if err := r.Get(ctx, req.NamespacedName, obj); err != nil {
+		logger.Info(req.NamespacedName.String() + " got deleted!")
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if r.ShouldSkip(ctx, obj) {
+		return ctrl.Result{}, nil
+	}
+
+	// ServerSideApply requires nil ManagedFields
+	obj.SetManagedFields(nil)
+
+	status := obj.GetStatus()
+
+	if !obj.GetDeletionTimestamp().IsZero() && obj.GetStatus().State != StateDeleting {
+		obj.SetStatus(status.WithState(StateDeleting).WithOperation("switched to deleting"))
+		return r.ssaStatus(ctx, obj)
+	}
+
+	// add finalizer
+	if controllerutil.AddFinalizer(obj, r.Finalizer) {
+		return r.ssa(ctx, obj)
+	}
+
+	var (
+		resourceCondition = metav1.Condition{
+			Type:               string(ConditionTypeResources),
+			Reason:             string(ConditionReasonResourcesAreAvailable),
+			Status:             metav1.ConditionFalse,
+			Message:            "resources are parsed and ready for use",
+			ObservedGeneration: obj.GetGeneration(),
+		}
+		installationCondition = metav1.Condition{
+			Type:               string(ConditionTypeInstallation),
+			Reason:             string(ConditionReasonReady),
+			Status:             metav1.ConditionFalse,
+			Message:            "installation is ready and resources can be used",
+			ObservedGeneration: obj.GetGeneration(),
+		}
+	)
+
+	// Processing
+	if status.State == "" {
+		presetConditions := 2
+		status.Conditions = make([]metav1.Condition, 0, presetConditions)
+		meta.SetStatusCondition(&status.Conditions, resourceCondition)
+		meta.SetStatusCondition(&status.Conditions, installationCondition)
+		status.Synced = make([]Resource, 0)
+		obj.SetStatus(status.WithState(StateProcessing).WithOperation("switched to processing"))
+		return r.ssaStatus(ctx, obj)
+	}
+
+	spec, err := r.Spec(ctx, obj)
+	if err != nil {
+		r.Event(obj, "Warning", "Spec", err.Error())
+		obj.SetStatus(status.WithState(StateError).WithErr(err))
+		return r.ssaStatus(ctx, obj)
+	}
+
+	skr, err := r.initializeClient(ctx, obj, spec)
+	if err != nil {
+		r.Event(obj, "Warning", "ClientInitialization", err.Error())
+		obj.SetStatus(status.WithState(StateError).WithErr(err))
+		return r.ssaStatus(ctx, obj)
+	}
+	resourceConverter := NewResourceConverter(skr, r.Namespace)
+
+	var renderer Renderer
+
+	switch spec.Mode {
+	case RenderModeHelm:
+		renderer = NewHelmRenderer(spec, skr, r.Options)
+		renderer = WrapWithRendererCache(renderer, spec, r.Options)
```

Comment:
IMO, this `WrapWithRendererCache` can put outside of switch case, since internally it will use `options` to decide if necessary to wrap, which also fit for RawRenderer

Reply:
For RawRenderer NoCache is not automatically set. Thus by default it will cache the raw manifest. I wanted to avoid this and I think it is still explicit enough. WDYT?

Reply:
but in this case, can we set `r.Options.ManifestCache` to `NoManifestCache` in case `RenderModeRaw`, then handle `WrapWithRendererCache` together after switch?

Reply:
Yes, I thought about this. I decided against this as I wanted to not mutate Options after initialising them. If you prefer a hardcoded version, of course I can do that. WDYT?

Reply:
ah, now I got it